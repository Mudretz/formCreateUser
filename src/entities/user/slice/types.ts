import { IOption } from "@src/components/common/CustomSelect/CustomSelect";

export interface UserState {
    contacts: Contacts;
    userInfo: UserInfo;
    advantages: Advantages;
    aboutMe: string;
}

export interface Contacts {
    phone: string;
    email: string;
}

export interface UserInfo {
    nickname: string;
    name: string;
    surname: string;
    sex: IOption | null;
}

export interface Advantages {
    advantages: string[];
    checkboxes: number[];
    radioOption: string;
}
