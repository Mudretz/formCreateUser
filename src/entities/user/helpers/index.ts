import { CreateUserRequest } from "../service/user.types";
import { UserState } from "../slice/user.types";

export const transformUserDataForApi = (data: UserState): CreateUserRequest => {
    const userInfo = {
        ...data.userInfo,
        sex: data.userInfo.sex?.label as string,
    };
    const userAdvantages = {
        ...data.userAdvantages,
        radioOption: data.userAdvantages.radioOption as number,
    };
    return {
        ...data.userContacts,
        ...userInfo,
        ...userAdvantages,
        ...data.userAbout,
    };
};
