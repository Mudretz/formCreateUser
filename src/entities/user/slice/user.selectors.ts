import { RootState } from "@src/app/store";

export const getUserContacts = (state: RootState) => {
    return state.user.userContacts;
};

export const getUserInfo = (state: RootState) => {
    return state.user.userInfo;
};

export const getUserAdvantages = (state: RootState) => {
    return state.user.userAdvantages;
};

export const getUserAbout = (state: RootState) => {
    return state.user.userAbout;
};
