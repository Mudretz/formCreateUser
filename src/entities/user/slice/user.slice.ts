import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import {
    UserAdvantages,
    UserContacts,
    UserInfo,
    UserState,
} from "./user.types";

const initialState: UserState = {
    userContacts: {
        phone: "",
        email: "",
    },
    userInfo: {
        nickname: "",
        name: "",
        surname: "",
        sex: null,
    },
    userAdvantages: {
        advantages: [
            {
                advantage: "",
            },
        ],
        checkboxes: [1],
        radioOption: "12",
    },
    userAbout: "",
};

const userSlice = createSlice({
    name: "userSlice",
    initialState,
    reducers: {
        userContactsReceived: (state, action: PayloadAction<UserContacts>) => {
            state.userContacts = action.payload;
        },
        userInfoReceived: (state, action: PayloadAction<UserInfo>) => {
            state.userInfo = action.payload;
        },
        userAdvantagesReceived: (
            state,
            action: PayloadAction<UserAdvantages>,
        ) => {
            state.userAdvantages = action.payload;
        },
        userAboutReceived: (state, action: PayloadAction<string>) => {
            state.userAbout = action.payload;
        },
    },
});

export const {
    reducer: userReducer,
    actions: {
        userContactsReceived,
        userInfoReceived,
        userAdvantagesReceived,
        userAboutReceived,
    },
} = userSlice;
