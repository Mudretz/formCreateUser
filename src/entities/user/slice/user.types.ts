import { IOption } from "@src/components/common/CustomSelect/CustomSelect";

export interface UserState {
    userContacts: UserContacts;
    userInfo: UserInfo;
    userAdvantages: UserAdvantages;
    userAbout: string;
}

export interface UserContacts {
    phone: string;
    email: string;
}

export interface UserInfo {
    nickname: string;
    name: string;
    surname: string;
    sex: IOption | null;
}

export interface UserAdvantages {
    advantages: string[];
    checkboxes: number[];
    radioOption: string;
}
