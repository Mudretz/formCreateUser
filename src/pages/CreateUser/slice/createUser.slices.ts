import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { CreateUserState } from "./createUser.types";

const initialState: CreateUserState = {
    activeStep: 1
};

const createUserSlice = createSlice({
    name: "createUser",
    initialState,
    reducers: {
        stepReceived: (state, action: PayloadAction<number>) => {
            state.activeStep = action.payload
        },
        stepIncrease: (state) => {
            state.activeStep += 1;
        },
        stepDecrease: (state) => {
            state.activeStep -= 1;
        }
    }
});

const { reducer: createUserReducer, actions } = createUserSlice;
export const { stepReceived, stepIncrease, stepDecrease } = actions;

export default createUserReducer;

