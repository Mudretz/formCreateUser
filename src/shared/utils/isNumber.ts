export const isNumber = (value: string) => {
    return !Number.isNaN(Number(value));
};
