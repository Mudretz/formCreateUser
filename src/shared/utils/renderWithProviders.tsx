import { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import setupStore, { AppStore, RootState } from "@src/app/store";
import { Provider } from "react-redux";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {},
) {
    function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
        return <Provider store={store}>{children}</Provider>;
    }
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
