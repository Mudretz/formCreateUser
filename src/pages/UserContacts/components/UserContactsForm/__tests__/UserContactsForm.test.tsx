import { screen } from "@testing-library/react";
import { UserContactsForm } from "..";
import { MemoryRouter } from "react-router-dom";
import setupStore from "@src/app/store";
import userEvent from "@testing-library/user-event";
import { renderWithProviders } from "@src/shared/utils/renderWithProviders";
import * as actions from "@src/entities/user/slice/user.slice";

describe("UserContact component", () => {
    test("Тестирование структуры UserContacts", async () => {
        const component = renderWithProviders(
            <MemoryRouter>
                <UserContactsForm />
            </MemoryRouter>,
        );
        expect(component).toMatchSnapshot();
    });
    // test("Тестирование скрина UserContact", async () => {
    //     const browser = await puppeteer.launch();
    //     const page = await browser.newPage();
    //     await page.goto("http://localhost:3000/");
    //     await page.waitForSelector("#main");
    //     const image = await page.screenshot();
    //     expect(image).toMatchImageSnapshot();
    //     page.close();
    // });
    test("Тестирование формы", async () => {
        const userContactPost = vi.spyOn(actions, "userContactsReceived");
        const store = setupStore();
        renderWithProviders(
            <MemoryRouter>
                <UserContactsForm />
            </MemoryRouter>,
            {
                store,
            },
        );

        const inputPhone = screen.getByTestId("input-phone");
        expect(inputPhone).toBeInTheDocument();
        await userEvent.type(inputPhone, "9999999999");
        expect(inputPhone).toHaveDisplayValue("+7 (999) 999-99-99");

        const emailInput = screen.getByTestId("input-email");
        expect(emailInput).toBeInTheDocument();
        await userEvent.type(emailInput, "test@gmail.com");
        expect(emailInput).toHaveDisplayValue("test@gmail.com");

        const btn = screen.getByRole("button");
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        expect(userContactPost).toHaveBeenCalledTimes(1);
        expect(userContactPost).toHaveBeenCalledWith({
            phone: "+7 (999) 999-99-99",
            email: "test@gmail.com",
        });

        expect(store.getState().user.userContacts).toEqual({
            phone: "+7 (999) 999-99-99",
            email: "test@gmail.com",
        });
    });
});
