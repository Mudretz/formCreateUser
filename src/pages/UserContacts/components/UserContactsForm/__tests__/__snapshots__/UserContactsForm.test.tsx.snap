// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserContact component > Тестирование структуры UserContacts 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="_main_376b55"
        id="main"
      >
        <form
          class="_form_376b55"
        >
          <div
            class="_form_67f839"
          >
            <label
              for="phone"
            >
              Номер телефона
            </label>
            <input
              class="_input_a02559"
              data-testid="input-phone"
              id="phone"
              name="phone"
              placeholder="+7 999 999-99-99"
              value=""
            />
            <p
              class="_error_67f839"
            >
               
            </p>
          </div>
          <div
            class="_form_bc7e7f"
          >
            <input
              class="_input_c0d863"
              data-testid="input-email"
              id="email"
              name="email"
              placeholder="tim.jennings@example.com"
            />
            <p
              class="_error_bc7e7f"
            >
               
            </p>
          </div>
          <button
            class="_button_376b55 _button_1ee4c7 _primary_1ee4c7"
            data-testid="button-start"
            id=":r1:"
            type="submit"
          >
            Начать
          </button>
        </form>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="_main_376b55"
      id="main"
    >
      <form
        class="_form_376b55"
      >
        <div
          class="_form_67f839"
        >
          <label
            for="phone"
          >
            Номер телефона
          </label>
          <input
            class="_input_a02559"
            data-testid="input-phone"
            id="phone"
            name="phone"
            placeholder="+7 999 999-99-99"
            value=""
          />
          <p
            class="_error_67f839"
          >
             
          </p>
        </div>
        <div
          class="_form_bc7e7f"
        >
          <input
            class="_input_c0d863"
            data-testid="input-email"
            id="email"
            name="email"
            placeholder="tim.jennings@example.com"
          />
          <p
            class="_error_bc7e7f"
          >
             
          </p>
        </div>
        <button
          class="_button_376b55 _button_1ee4c7 _primary_1ee4c7"
          data-testid="button-start"
          id=":r1:"
          type="submit"
        >
          Начать
        </button>
      </form>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
